File: Solutions/python/yacht/yacht.py
# Score categories.
# Change the values as you see fit.
YACHT = "yacht"
ONES = "ones"
TWOS = "twos"
THREES = "threes"
FOURS = "fours"
FIVES = "fives"
SIXES = "sixes"
FULL_HOUSE = "full_house"
FOUR_OF_A_KIND = "four_of_a_kind"
LITTLE_STRAIGHT = "little_straight"
BIG_STRAIGHT = "big_straight"
CHOICE = "choice"


def score(dice, category):
    number_map = {"ones": 1, "twos": 2, "threes": 3, "fours": 4, "fives": 5, "sixes": 6}

    if category in number_map:
        return dice.count(number_map[category]) * number_map[category]

    if category == "yacht":
        return 50 if len(set(dice)) == 1 else 0

    if category == "full_house":
        counts = set(dice.count(die) for die in dice)
        return sum(dice) if counts == {2, 3} else 0

    if category == "four_of_a_kind":
        for die in set(dice):
            if dice.count(die) >= 4:
                return die * 4
        return 0

    if category == "little_straight":
        return 30 if sorted(dice) == [1, 2, 3, 4, 5] else 0

    if category == "big_straight":
        return 30 if sorted(dice) == [2, 3, 4, 5, 6] else 0

    if category == "choice":
        return sum(dice)

    return 0

File: Solutions/elixir/yacht/lib/yacht.ex
defmodule Yacht do
  @type category ::
          :ones
          | :twos
          | :threes
          | :fours
          | :fives
          | :sixes
          | :full_house
          | :four_of_a_kind
          | :little_straight
          | :big_straight
          | :choice
          | :yacht

  @doc """
  Calculate the score of 5 dice using the given category's scoring method.
  """
  @spec score(category :: category(), dice :: [integer]) :: integer
  def score(category, dice) do
    case category do
      :ones -> sum_for_number(dice, 1)
      :twos -> sum_for_number(dice, 2)
      :threes -> sum_for_number(dice, 3)
      :fours -> sum_for_number(dice, 4)
      :fives -> sum_for_number(dice, 5)
      :sixes -> sum_for_number(dice, 6)
      :full_house -> full_house_score(dice)
      :four_of_a_kind -> four_of_a_kind_score(dice)
      :little_straight -> if Enum.sort(dice) == [1, 2, 3, 4, 5], do: 30, else: 0
      :big_straight -> if Enum.sort(dice) == [2, 3, 4, 5, 6], do: 30, else: 0
      :choice -> Enum.sum(dice)
      :yacht -> if Enum.uniq(dice) == [hd(dice)], do: 50, else: 0
    end
  end

  @doc false
  defp sum_for_number(dice, number) do
    dice
    |> Enum.filter(&(&1 == number))
    |> Enum.sum()
  end

  @doc false
  defp full_house_score(dice) do
    counts = dice |> Enum.frequencies() |> Map.values() |> Enum.sort()
    if counts == [2, 3], do: Enum.sum(dice), else: 0
  end

  @doc false
  defp four_of_a_kind_score(dice) do
    counts = Enum.frequencies(dice)

    case Enum.find_value(counts, fn {_k, v} -> v >= 4 end) do
      nil -> 0
      _ -> dice |> Enum.filter(fn x -> counts[x] >= 4 end) |> Enum.take(4) |> Enum.sum()
    end
  end
end
